io-engine/src/core/reactor.rs

```diff
@@ -624,3 +635,101 @@ impl Future for &'static Reactor {
         }
     }
 }
+
+/// Maximum number of heartbeats a reactor is allowed to miss
+/// before it is classified as frozen.
+#[cfg(feature = "diagnostics")]
+const MAX_MISSED_HEARTBEATS: u64 = 3;
+
+/// Monitor health for all reactors: all available reactors are constantly
+/// monitored for liveness.
+#[cfg(feature = "diagnostics")]
+pub async fn reactor_monitor_loop() {
+    use std::sync::atomic::{AtomicU64, Ordering};
+
+    /// Metadata for every reactor being monitored by the reactor monitor.
+    struct ReactorRecord {
+        frozen: bool,
+        reactor: &'static Reactor,
+        reactor_tick: &'static AtomicU64,
+        core: u32,
+    }
+
+    let mut interval = tokio::time::interval(Duration::from_secs(1));
+    let mut tick: u64 = 0;
+    let mut reactor_state: Vec<ReactorRecord> =
+        Vec::with_capacity(Cores::count().id() as usize);
+    static REACTOR_TICKS: OnceCell<Vec<AtomicU64>> = OnceCell::new();
+
+    info!(
+        cores = Cores::count().id(),
+        "Starting reactor health monitor loop"
+    );
+
+    // Intialize shared counters for heartbeat futures sent to reactors.
+    let heartbeat_ticks = REACTOR_TICKS.get_or_init(|| {
+        std::iter::repeat_with(|| AtomicU64::new(0))
+            .take(Cores::count().id() as usize)
+            .collect::<Vec<AtomicU64>>()
+    });
+
+    // Initialize reactor records.
+    for (id, core) in Cores::count().into_iter().enumerate() {
+        let reactor = Reactors::get_by_core(core)
+            .unwrap_or_else(|| panic!("Can't get reactor for core {}", core));
+        let reactor_tick =
+            heartbeat_ticks.get(id).expect("Failed to get tick item");
+
+        reactor_state.push(ReactorRecord {
+            frozen: false,
+            reactor,
+            reactor_tick,
+            core,
+        });
+    }
+
+    loop {
+        // Schedule heartbeat futures on every reactor, ignoring reactors
+        // which are already frozen.
+        for (id, r) in reactor_state.iter().enumerate() {
+            // For frozen reactors there are already N scheduled heartbeat
+            // futures that haven't resolved yet, so maintain exactly this delta
+            // by just adjusting the tick counter.
+            if r.frozen {
+                heartbeat_ticks[id].fetch_add(1, Ordering::Relaxed);
+            } else {
+                // Send heartbeat future to the reactor.
+                r.reactor.send_future(async move {
+                    heartbeat_ticks[id].fetch_add(1, Ordering::Relaxed);
+                });
+            }
+        }
+
+        // Wait till heartbeat check interval elapses and check ticks
+        // reported by every reactor.
+        interval.tick().await;
+        tick += 1;
+
+        for r in &mut reactor_state {
+            if r.frozen {
```

Comment:
Should we "send" this knowledge to the registration subsystem? Would that help? Eg: if we know a reactor is stuck should the control-plane stop trying to send grpc to this node? Atm the registration keep-alive keep being sent so the control-plane thinks all is good.